### ISSUE TITLE
RST_STREAM takes time

### ISSUE THREAD

**Issue by minhkstn** [nghttp2/nghttp2#675]
When I use RST_SREAM, it's good to cancel the stream. However, if I wanna cancel all the PUSH_PROMISE_IDs that haven't pushed to the client, then the server needs time (from 200ms to 1000ms) to do this. How can I make it cancel immediately?

**Comment by tatsuhiro-t** [comment]
I'd like to know how you measure the time?   Any source code to reproduce this issue?

**Comment by MisterduncanPr** [comment]
I measure time as follow this:
std::chrono::duration<double, std::milli> time=req->timing.response_end_time - req->timing.request_start_time;

I got this trouble when I don't want to receive some push_promise.promised_stream_ids (eg. I cancelled stream_id from 50 to 100). 
 
I don't know how server will adapt after client send a message: nghttp2_submit_rst_stream(session, NGHTTP2_FLAG_NONE, stream_id ,NGHTTP2_CANCEL);

**Comment by tatsuhiro-t** [comment]
So you measured time in nghttp?
Are you sending RST_STREAM from client from nghttp client?
Are you sure that you are using promised_stream_id instead of stream_id of PUSH_PROMISE?
Do you have source code or patch to demonstrate this thing?
It takes some time for server to cancel response because of latency.

**Comment by MisterduncanPr** [comment]
Yes, I am sending RST_STREAM from nghttp client and measured time in nghttp, too.
This is my source code.
if( rst_stream =="1"){
          for(int i=number*2*session_id+2;i<=number*2*session_id+number*2;i=i+2){
	  nghttp2_submit_rst_stream(session, NGHTTP2_FLAG_NONE,i,NGHTTP2_CANCEL);
}
  }
Are there any differences when I use promised_stream_id or stream_id of PUSH_PROMISE?

**Comment by tatsuhiro-t** [comment]
> Are there any differences when I use promised_stream_id or stream_id of PUSH_PROMISE?

stream_id of PUSH_PROMISE is an associated stream, which is not pushed stream.
RST_STREAMing such stream ID does not close pushed stream.

nghttp tends to read all data without sending frames especially when you are running client and server in the same machine.  Perhaps, nghttp2_session_send call may be delayed in these cases.  Are you running client and server in the same machine?

**Comment by tatsuhiro-t** [comment]
... especially if you use -v option in nghttp, which could delay nghttp2_session_send call.

**Comment by MisterduncanPr** [comment]
Thank you. Yes, I am running in the same machine.
So, do I have to RST_STREAM  promised_stream_id to terminate steam, don't I? And maybe I will try running in different machine.

**Comment by MisterduncanPr** [comment]
Sorry, I am running client and server in 2 different virtual machines using vmware workstation.
Which does statement or function help me terminate the pushing stream from nghttp client? And If so, What will server adaptive when it receive that statement/function?

**Comment by tatsuhiro-t** [comment]
Use of nghttp2_submit_rst_stream is correct approach to cancel stream.
When you cancel pushed stream when you get PUSH_PROMISE, use nghttp2_frame.push_promise.promised_stream_id to cancel stream.
For other frames (DATA or HEADERS), use nghttp2_frame.hd.stream_id.

I don't know which server you are using, but nghttp2 implementation stops sending DATA or HEADERS right away after it receives RST_STREAM.
Be aware that after sending RST_STREAM, and before server receives it, server may already commit HEADERS or DATA to network stack, which cannot be undone.

**Comment by haidang12** [comment]
Is there any statement or function help me terminate all streams which are not push in a session? Thank you

**Comment by tatsuhiro-t** [comment]
Use nghttp2_submit_rst_stream to cancel stream.  You can pass any existing stream ID there.

**Comment by rodrigobdz** [comment]
@haidang12 how do you cancel all the PUSH_PROMISE_IDs that haven't pushed to the client?
In your example code you use this loop `for(int i=number*2*session_id+2;i<=number*2*session_id+number*2;i=i+2){}`
Could you explain it a little bit further?
